sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/DataflowPipelineRunner.java

```diff
@@ -601,17 +603,35 @@ public DataflowPipelineJob run(Pipeline pipeline) {
     }
 
     if (!Strings.isNullOrEmpty(options.getDataflowJobFile())) {
-      try (PrintWriter printWriter = new PrintWriter(
-          new File(options.getDataflowJobFile()))) {
+      try {
+        WritableByteChannel writer =
+            IOChannelUtils.create(options.getDataflowJobFile(), MimeTypes.TEXT);
+        PrintWriter printWriter = new PrintWriter(Channels.newOutputStream(writer));
         String workSpecJson = DataflowPipelineTranslator.jobToString(newJob);
         printWriter.print(workSpecJson);
-        LOG.info("Printed workflow specification to {}", options.getDataflowJobFile());
+        printWriter.flush();
+        printWriter.close();
+        LOG.info("Printed job specification to {}", options.getDataflowJobFile());
       } catch (IllegalStateException ex) {
-        LOG.warn("Cannot translate workflow spec to json for debug.");
-      } catch (FileNotFoundException ex) {
-        LOG.warn("Cannot create workflow spec output file.");
+        String error = "Cannot translate workflow spec to JSON.";
+        if (hooks != null && hooks.failOnJobFileWriteFailure()) {
+          throw new RuntimeException(error, ex);
+        } else {
+          LOG.warn(error);
+        }
+      } catch (IOException ex) {
+        String error =
+            String.format("Cannot create output file at {}", options.getDataflowJobFile());
+        if (hooks != null && hooks.failOnJobFileWriteFailure()) {
```

Comment:
It may be worth putting this logic (handling hooks, throwing or not, etc.) into a helper method?